Create BCFtools_stats_plot subworkflow

yz533cb opened the pull request:
<!--
# sanger-tol/variantcomposition pull request

Many thanks for contributing to sanger-tol/variantcomposition!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/sanger-tol/variantcomposition/tree/main/.github/CONTRIBUTING.md)
-->

Add a QC feature using `BCFtools-stats` and `plot-vcfstats`. The feature is in a subworkflow.
Modules `BCFtools-stats` and `BCFtools-ROH` from the two subworkflows both need an index file. Therefore, `TABIX` is moved to the main workflow.

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/sanger-tol/variantcomposition/tree/main/.github/CONTRIBUTING.md)
- [x] Make sure your code lints (`nf-core pipelines lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

yz533cb committed:
Add BCFtools_stats module

yz533cb committed:
Update pipeline diagram

yz533cb committed:
Add plot_vcfstats module (conda profile can run, output empty; container cannot run)

yz533cb committed:
Fix empty output; local test and test_full pass with Conda and Docker

yz533cb committed:
Fix matplotlib cache when using Singularity

yz533cb committed:
Create BCFtools_stats_plot subworkflow; move TABIX to main workflow

sanger-tolsoft commented:
<!-- add-pr-comment:add-pr-comment -->

> [!WARNING]
> Newer version of the nf-core template is available.
>
> Your pipeline is using an old version of the nf-core template: 3.3.1.
> Please update your pipeline to the latest version.
>
> For more documentation on how to update your pipeline, please see the [nf-core documentation](https://github.com/nf-core/tools?tab=readme-ov-file#sync-a-pipeline-with-the-template) and [Synchronisation documentation](https://nf-co.re/docs/contributing/sync).

yz533cb committed:
Ignore PDF output in nf-test

yz533cb committed:
Update BCFtools-stats module

yz533cb committed:
Move features subworkflow to a folder

yz533cb committed:
Update pipeline diagram

yz533cb committed:
Update CHANGELOG

yz533cb committed:
Update README

yz533cb committed:
Update output.md

yz533cb committed:
Remove 'nf-core' string after 'TODO' string to avoid linting warnings

yz533cb committed:
Temporary remove the 'GitHub Actions CI Status' in README, update ro-crate-metadata (can recover in the next template-update)

yz533cb committed:
Merge branch 'dev' into QC_by_BCFtools_stats

yz533cb committed:
Fix md5 of BCFtools-RoH gVCF output that was missed when solving merge conflict

yz533cb committed:
Add Matplotlib and Tectonic to CHANGELOG

yz533cb committed:
Ignore log file in nf-test

yz533cb commented:
When testing with Singularity, an error occurred:
```
Command error:
  Parsing bcftools stats output: GCA_947369205.1_OX376310.1_CANBKR010000003.1.pacbio.icCanRufa1_deepvariant.g.vcf.bcftools_stats.txt
  Plotting graphs: python3 plot.py
  Fontconfig error: No writable cache directories
  Creating PDF: tectonic summary.tex >plot-vcfstats.log 2>&1
  The command exited with non-zero status, please consult the output of tectonic: GCA_947369205.1_OX376310.1_CANBKR010000003.1.pacbio.icCanRufa1_deepvariant.g.vcf_plots/plot-vcfstats.log
```
It's because that Nextflow 23.10 adds the `--no-home` option when using Singularity. So `matplotlib` failed to write cache files that are needed by `tectonic`. There's a good discussion [here](https://github.com/nf-core/ampliseq/issues/654).

Current solution: creating a `HOME` folder for cache files by adding these lines to the module script:
```
mkdir nxf_home
export HOME=\$PWD/nxf_home
```

yz533cb commented:
`nf-test` fails because the PDF, log, and tex output differs between runs, while the raw data files (`.txt`, `.dat`, `.png`) remain identical.

- **PDF files**: the differences are likely due to embedded metadata (timestamps, IDs, etc.), so I added PDFs to `tests/.nftignore`

- **Log files**: the `plot-vcfstats.log` file shows occasional differences. These are rare. I identified the problem when run `--update-snapshot` multiple times and check the changes in snapshot file in VSCode. But when compared the files from 20 different runs, I didn't discover the difference

Comparing the log between `test` and `test_full` profiles, I found the difference in an additional warning at line 127:
```
warning: summary.tex:115: Overfull \vbox (19.15474pt too high) has occurred while \output is active
```
So now I added log files to `tests/.nftignore` as well.

- **Conda vs Singularity**
  - with Conda, `tectonic` produces two extra files `summary.aux` and `summary.log`
    - with Conda, `plot-vcfstats.log` is identical with `summary.log`, which is different from the `plot-vcfstats.log` with Singularity
    - this may be due to differences in how `techtonic` is packaged/built in Conda vs container?

  - the `summary.tex` file also differs:

Conda (line 5):
```
pdflatex summary.tex
```

Singularity (line 5):
```
tectonic summary.tex
```
This shows a difference in how the documentation is generated.

**Conclusion**: I've added `log`, `aux`, `tex` and `pdf` in `test/.nftignore`; and added `log`, `aux` in `test/default.nf.test`:
```
stable_name = getAllFilesFromDir( ignore: [ 'plot_vcfstats/*/*.{aux,log}' ] )
```
so that `nf-test` only snapshots the deterministic data outputs.

yz533cb committed:
Ignore aug and log files generated by Conda-pdflatex in nf-test

muffato reviewed (commented):
The code itself is fine. It's a clean addition

Looking at the pipeline outputs, I would propose to make a few changes:

1. If the `bcftools_stats.txt` file itself can get big, it should be compressed before copying to the output directory
2. I would put `plot-vcfstats.pdf` in the main output directory directly
3. I would compress the `vcf_plots` directory and put the archive in the main output directory as well.

yz533cb committed:
Tar and compress plot-vcfstats output; output move to main outdir

yz533cb commented:
> The code itself is fine. It's a clean addition
> 
> Looking at the pipeline outputs, I would propose to make a few changes:
> 
> 1. If the `bcftools_stats.txt` file itself can get big, it should be compressed before copying to the output directory
> 2. I would put `plot-vcfstats.pdf` in the main output directory directly
> 3. I would compress the `vcf_plots` directory and put the archive in the main output directory as well.

@muffato Thank you for the review! 

The `bcftools_stats.txt` file is about 20-30 KB for single-VCF files and 1-2 MB for multi-VCF files. The file is now compressed. (edited after adding compression)

Great idea to compress the `vcf_plots` folder to a whole. It’s now archived and compressed using `tar --gzip` via the `tar` module.

Now all outputs are written to the main outdir.

yz533cb committed:
Compress bcftools_stats.txt; update output.md

muffato reviewed (commented):
Almost almost there :) The only thing really is that it's weird to use BGZIP in this case.
See the manual page: https://www.htslib.org/doc/bgzip.html
> Bgzip compresses files in a similar manner to, and compatible with, gzip(1). The file is compressed into a series of small (less than 64K) 'BGZF' blocks. **This allows indexes to be** built against the compressed file and **used to retrieve portions of the data without having to decompress the entire file.**

It's the perfect tool for VCF files because they can be indexed by coordinates, and then we can retrieve the variants in one region with decompressing just the relevant part of the file. Really useful for large files.
The bcftools stats file doesn't have this kind of index format so there will be no benefit over a regular GZIP.

yz533cb commented:
> Almost almost there :) The only thing really is that it's weird to use BGZIP in this case. See the manual page: https://www.htslib.org/doc/bgzip.html
> 
> > Bgzip compresses files in a similar manner to, and compatible with, gzip(1). The file is compressed into a series of small (less than 64K) 'BGZF' blocks. **This allows indexes to be** built against the compressed file and **used to retrieve portions of the data without having to decompress the entire file.**
> 
> It's the perfect tool for VCF files because they can be indexed by coordinates, and then we can retrieve the variants in one region with decompressing just the relevant part of the file. Really useful for large files. The bcftools stats file doesn't have this kind of index format so there will be no benefit over a regular GZIP.

Thanks for pointing it out! Yes yes, that makes sense. (I feel the need to understand more about the softwares and formats I've used so far..)

I did a quick search through nf-core modules and there isn’t a dedicated `gzip` module. Now we have several options:

1. Use the [pigz/compress](https://github.com/nf-core/modules/blob/master/modules/nf-core/pigz/compress/main.nf) module, which is a parallel implementation of gzip and produces standard `.gz` files

2. Adapt the existing [gunzip](https://github.com/nf-core/modules/blob/master/modules/nf-core/gunzip/main.nf) module, which uses `gzip -d`; maybe wen can remove the `-d` flag, add a comment to explain, and add a `diff` file

3. Write a new lightweight gzip module (could be just a small tweak of the gunzip one)

4. There's also a [zip](https://github.com/nf-core/modules/blob/master/modules/nf-core/zip/main.nf) module using 7z, which create a `.zip` file

5. Or we can use the tar module


Option 1 looks like a valid choice to me. What do you think?

muffato commented:
> I did a quick search through nf-core modules and there isn’t a dedicated `gzip` module. Now we have several options:
> 
> 1. Use the [pigz/compress](https://github.com/nf-core/modules/blob/master/modules/nf-core/pigz/compress/main.nf) module, which is a parallel implementation of gzip and produces standard `.gz` files
> (...)
> 
> Option 1 looks like a valid choice to me. What do you think?

pigz is fine 👍🏼

yz533cb committed:
Use Pigz to compress stats txt file

yz533cb committed:
Update CHANGELOG
